kedro/runner/runner.py

```diff
@@ -94,9 +94,16 @@ def run(
                 f"Pipeline input(s) {unsatisfied} not found in the DataCatalog"
             )
 
+        # Identify MemoryDataset in the catalog
+        memory_datasets = {
+            ds_name
+            for ds_name, ds in catalog._datasets.items()
+            if isinstance(ds, MemoryDataset)
+        }
+
         # Check if there's any output datasets that aren't in the catalog and don't match a pattern
-        # in the catalog.
-        free_outputs = pipeline.outputs() - set(registered_ds)
+        # in the catalog and include MemoryDataset.
+        free_outputs = pipeline.outputs() - (set(registered_ds) - memory_datasets)
```

Comment:
nit: The definition of `free_outputs` has always been confusing to me, I think what are returned here is "in_memory_dataset" as we are trying to return the dataset as long as there are no I/O penalties.

Feel free to come up with other names.